### PR TITLE
Initial fixes for IO.read and check body length per https://github.com/acquia/http-hmac-spec/pull/12

### PR DESCRIPTION
**Pull request by pwolanin**

### CODE REVIEW

File: lib/acquia-http-hmac/rack_authenticate.rb [hunk 1]
```diff
@@ -89,11 +89,11 @@ def args_for_authenticator(env, attributes)
 
       def valid_body?(env)
         request = Rack::Request.new(env)
-        if ['GET', 'HEAD'].include?(request.request_method)
+        body = request.body.read   # read the incoming request IO stream
```
Comment (glennpratt) [inline lib/acquia-http-hmac/rack_authenticate.rb]: I'd like to see `request.body.rewind` after this. The Rack interface requires `#rewind` be supported and I'm guessing other users don't call it in advance.

Comment (glennpratt) [inline lib/acquia-http-hmac/rack_authenticate.rb]: I noticed empty bodies in another middleware after not calling rewind.

**Comment by pwolanin** [comment]
@glennpratt added a rewind

**Comment by glennpratt** [comment]
@pwolanin looking at rails, they always guard the rewind call, so maybe that is a good idea, even though Rack requires it.

```ruby
request.body.rewind if request.body.respond_to?(:rewind)
```
https://github.com/rails/rails/blob/d9bdb61158812484aabf2a31fb943dccb7749887/actionpack/lib/action_dispatch/http/request.rb#L300

http://www.rubydoc.info/github/rack/rack/Rack/RewindableInput

File: lib/acquia-http-hmac/rack_authenticate.rb [hunk 2]
```diff
@@ -89,11 +89,14 @@ def args_for_authenticator(env, attributes)
 
       def valid_body?(env)
         request = Rack::Request.new(env)
-        if ['GET', 'HEAD'].include?(request.request_method)
+        # Read the incoming request IO stream.
+        body = request.body.read
+        # Allow other middleware to access the body also.
+        request.body.rewind
+        if (body.length == 0) && env['HTTP_X_AUTHORIZATION_CONTENT_SHA256'].nil?
```
Comment (glennpratt) [inline lib/acquia-http-hmac/rack_authenticate.rb]: Nit: `body.empty?`

**Comment by glennpratt** [comment]
Looks good to me.

**Comment by jamesiarmes** [comment]
:+1:

**Pull request merged**